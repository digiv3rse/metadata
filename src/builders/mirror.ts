import { v4 } from 'uuid';

import { evaluate } from './ValidationError';
import { RecursiveUnbrand } from './utils';
import {
  MirrorMetadata,
  MirrorMetadataSchema,
  MirrorSchemaId,
  MirrorMetadataDetails,
} from '../publication';
import { Prettify } from '../utils.js';

/**
 * @private
 * @privateRemarks MUST stay very @private to produce usable docs
 */
type MirrorDetails = Prettify<RecursiveUnbrand<Omit<MirrorMetadataDetails, 'id'>>>;
/**
 * All {@link MirrorMetadataDetails} fields with:
 * - `id` defaults to a UUID
 */
export type MirrorOptions = MirrorDetails & {
  /**
   * A unique identifier that in storages like IPFS ensures the uniqueness of the metadata URI.
   *
   * @defaultValue a UUID
   */
  id?: string;
};
/**
 * Creates a valid MirrorMetadata.
 *
 * @category Compose
 * @param input - Use your IDE suggestions for an enhanced development experience
 *
 * @example
 * ```ts
 * const metadata = mirror({
 *   appId: 'com.example.app',
 * });
 * ```
 */
export function mirror({ id = v4(), ...others }: MirrorOptions): MirrorMetadata {
  return evaluate(
    MirrorMetadataSchema.safeParse({
      $schema: MirrorSchemaId.LATEST,
      digi: {
        id,
        ...others,
      },
    }),
  );
}
