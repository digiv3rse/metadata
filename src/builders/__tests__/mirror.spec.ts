import { describe, expect, it } from '@jest/globals';

import { mirror } from '../mirror.js';

describe(`Given the "${mirror.name}" metadata builder`, () => {
  describe(`when using it`, () => {
    it('should return a valid MirrorMetadata', () => {
      const metadata = mirror({
        appId: 'com.example.app',
      });

      expect(metadata).toMatchSnapshot({
        digi: {
          id: expect.any(String),
        },
      });
    });
  });
});
