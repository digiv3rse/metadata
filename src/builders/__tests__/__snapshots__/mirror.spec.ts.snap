// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given the "mirror" metadata builder when using it should return a valid MirrorMetadata 1`] = `
{
  "$schema": "https://json-schemas.digiv3rse.xyz/publications/mirror/1.0.0.json",
  "digi": {
    "appId": "com.example.app",
    "id": Any<String>,
  },
}
`;
